### PR TITLE
feat(llm): replay-based harness unit tests for Claude Code + Codex

### PR DESCRIPTION
**Pull request by MatejKosec**
## Summary

Adds replay-based harness integration tests for Dynamo's Anthropic `/v1/messages` (Claude Code) and OpenAI `/v1/responses` (Codex) endpoints. A new `ReplayEngine` implements `AsyncEngine` by replaying handwritten `chat.completion.chunk` JSONL through the real HTTP service; the emitted SSE / unary bodies are parsed, ID/timestamp-redacted, and snapshotted with `insta`.

All tests reuse existing dev-deps — **zero new dependencies** — and live alongside other integration tests under `lib/llm/tests/`.

## Relationship to other PRs

**Complementary to #8283** (end-to-end compliance CI for Responses). That PR pins the upstream OpenResponses bun suite + `codex exec` smoke against a real sglang backend; this PR adds the **fast unit-level tier** using `ReplayEngine` through `AsyncEngine`. Both are needed:

| Aspect | #8283 (compliance) | This PR (harness) |
|---|---|---|
| Tier | End-to-end (real sglang + GPU) | Unit (in-process HTTP + `ReplayEngine`) |
| Runtime | ~minutes / run, GPU required | ~seconds, no GPU |
| Catches | Wire-shape conformance vs upstream spec | Translation logic, SSE framing, stream-converter edge cases |
| Deps | `bun`, `codex` CLI, sglang container | Only existing cargo dev-deps |
| Coverage | 6 pinned upstream tests | 19 scenarios: multi-turn reasoning, deep tool chains, multimodal, stop-sequence, structured output, cancellation |

Compliance catches wire conformance; harness catches translation logic and gives fast pre-GPU feedback.

Multimodal coverage here uses the **Anthropic** `content_block { type: "image", source }` shape; Responses-side multimodal (the new `InputImageContent { detail, file_id, image_url }` in #8283) is deferred to a follow-up that rebases onto #8283.

### Anthropic `/v1/messages` (10 tests, `anthropic_replay.rs`)

| ID | Name | Scenario |
|----|------|----------|
| A1 | `hello_unary` | Plain text, non-streaming `message` |
| A2 | `hello_streaming` | Plain text SSE with `message_start` -> `content_block_delta` -> `message_stop` |
| A3 | `tool_use_single` | Single `tool_use` block (`get_weather`) with partial-JSON input |
| A4 | `tool_use_parallel` | Two parallel `tool_use` blocks (weather + time) in one turn |
| A5 | `tool_result_roundtrip` | Multi-turn with `tool_result` in the user message |
| A6 | `cancel_midstream` | Client drops connection after first chunk -> engine observes cancellation |
| A7 | `count_tokens_shape` | `/v1/messages/count_tokens` response shape snapshot |
| A8 (extra) | `stop_sequence_streaming` | `stop_reason: stop_sequence` carried through the stream converter |
| A9 (extra) | `multimodal_image_input` | Image `content_block` on request side (unary response) |
| A10 (extra) | `multiturn_continuation` | User -> assistant -> user streaming continuation |

### OpenAI `/v1/responses` (9 tests, `responses_replay.rs`)

| ID | Name | Scenario |
|----|------|----------|
| R1 | `hello_streaming` | Plain text streaming via Responses |
| R2 | `reasoning_summary` | `reasoning.effort: low`, reasoning_content deltas |
| R3 | `function_call_single` | Single `function_call` output item |
| R4 | `function_call_parallel` | Two parallel `function_call`s in one turn |
| R5 | `function_call_output_followup` | Follow-up turn with `function_call_output` input |
| R6 | `cancel_midstream` | Client drops connection after first chunk -> engine observes cancellation |
| R7 (extra) | `deep_tool_chain` | 3 sequential function calls (`ls` -> `grep` -> `git_diff`) |
| R8 (extra) | `structured_output` | `text.format: json_schema` streaming of a JSON object |
| R9 (extra) | `multiturn_reasoning` | User/assistant/user continuation with reasoning content |

## Layout

```
lib/llm/tests/
├── common/
│   ├── harness.rs           # HarnessService builder, SSE parser, redactor
│   └── replay_engine.rs     # ReplayEngine: AsyncEngine<ChatCompletion> from JSONL
├── anthropic_replay.rs      # A1-A10 with temp_env + serial_test
├── responses_replay.rs      # R1-R9
├── data/replays/harness/    # 13 handwritten JSONL fixtures
└── snapshots/               # 17 insta snapshots
```

## Key design points

- **No env-var drift**: Anthropic is gated by `DYN_ENABLE_ANTHROPIC_API`; each A-test wraps in `temp_env::async_with_vars` and runs `#[serial]` so the process-global var doesn't race.
- **Cancellation detection**: `ReplayEngine::generate` spawns a detached watcher on `ctx.stopped()`; client disconnects cancel the response stream, which the watcher observes without needing the generator's stream body to survive.
- **Redaction**: `id`, `request_id`, `message_id`, `response_id`, `msg_id`, `item_id` -> `"[ID]"`; `created`/`*_at` -> `1_000_000_000`; `signature` -> `"[SIG]"`; `system_fingerprint` -> `"[FP]"`. Keeps snapshots stable across runs.
- **Fixture format**: plain JSONL with `//` comments permitted, one `NvCreateChatCompletionStreamResponse` per line.

## Test plan

- [x] `cargo test -p dynamo-llm --test anthropic_replay --test responses_replay` — 19/19 pass
- [x] Snapshot stability verified by re-running tests without `INSTA_UPDATE` — all 17 snapshots match
- [x] `cargo fmt -p dynamo-llm --check` clean
- [x] `cargo clippy -p dynamo-llm --tests -- -D warnings` clean
- [x] `cancel_midstream` reliably observes engine-side cancellation (`observed_cancels >= 1`) under 5s deadline

cc @ishandhanani for review — paired with #8283.

## Summary by CodeRabbit

* **Tests**
  * Expanded integration coverage for streaming chat and responses endpoints, including text replies, tool use, multi-turn conversations, structured output, multimodal input, stop sequences, and client disconnect handling.
  * Added replay-based scenarios to verify consistent behavior across unary and streamed responses.

* **Documentation**
  * Added guidance for maintaining replay fixtures used by the test harness.

### CODE REVIEW

File: lib/llm/tests/common/replay_engine.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+//! ReplayEngine: an `AsyncEngine` impl that replays pre-recorded
+//! `chat.completion.chunk` JSONL as a stream. Used by the harness-style
+//! integration tests under `lib/llm/tests/` to drive the Anthropic and
+//! Responses surfaces with deterministic backend output.
+//!
+//! The replay format is one `NvCreateChatCompletionStreamResponse` per line
+//! — the exact shape that an upstream engine (vLLM, sglang, TRT-LLM) would
+//! emit. Test fixtures are handwritten JSONL under `tests/data/replays/harness/`.
+
+#![allow(dead_code)]
```
Comment (rmccorm4) [inline lib/llm/tests/common/replay_engine.rs:13]: Why dead code?

File: lib/llm/tests/common/harness.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+//! Shared helpers for the Anthropic and Responses replay tests.
+
+#![allow(dead_code)]
```
Comment (rmccorm4) [inline lib/llm/tests/common/harness.rs:6]: Why dead code?

File: lib/llm/tests/common/replay_engine.rs
Comment (rmccorm4) [inline lib/llm/tests/common/replay_engine.rs:1]: Does it make sense to reuse existing `python -m dynamo.replay` or its underlying code instead?

File: lib/llm/tests/common/replay_engine.rs (hunk 1 above)
Comment (MatejKosec) [reply lib/llm/tests/common/replay_engine.rs]: Addressed in the current diff: the `#![allow(dead_code)]` is gone now that the helper code is wired into the replay tests.

File: lib/llm/tests/common/replay_engine.rs
Comment (MatejKosec) [reply lib/llm/tests/common/replay_engine.rs]: I kept this as a small Rust `AsyncEngine` rather than reusing `python -m dynamo.replay` because these are Rust HTTP-service tests that run in-process. They need direct access to the engine cancellation context, and depending on a separate Python replay process/deployment would make the tests heavier and less targeted. I added that rationale to the module docs in `replay_engine.rs`.

**Comment by MatejKosec** [comment]
@rmccorm4 on the 'why dead code' question: the helper functions (`observed_cancels`, `from_jsonl`) are part of the test harness API surface in `lib/llm/tests/common/` but not every helper is exercised by the current tests, which triggers Rust's dead-code lint. They're building blocks for future test cases that need cancel-count assertions or direct JSONL loading.

On reusing `python -m dynamo.replay`: explained in the module-level doc comment at the top of `replay_engine.rs` — these tests run the Rust HTTP service in-process and need direct access to the `CancellationToken` engine context, which a separate Python process can't provide. The small Rust `ReplayEngine` keeps everything in-process and deterministic, with no Python subprocess dependency.

**Comment by rmccorm4** [comment]
> on the 'why dead code' question

If we don't need the helpers yet, let's take them out of this PR please, we generally don't want to keep dead code around if we end up forgetting or deprioritizing the follow-ups that would use them later (which happens all the time). We can just have a small inline comment about helpers that may be helpful in future if that helps preserve some context for next time when it will be needed.

> On reusing python -m dynamo.replay

I'm not that familiar with the exact `replay` source code, but AFAIK it is largely written in rust (Replay and Mocker logic) - check with @PeaBrane on this. By `python -m dynamo.replay` I just mean the replay/mocker work Rudy and others are doing, not the specific python wrapper code there, it's fine if you need the rust hooks directly for a certain reason.

**Comment by MatejKosec** [comment]
@rmccorm4 I rechecked both points against the current branch.

1. The two helpers mentioned in the earlier reply are exercised by the tests rather than retained for future use. `ReplayEngine::from_jsonl` is called by every `HarnessService::start`, and `observed_cancels` is asserted by both the Anthropic and Responses cancellation tests.

2. I compared this harness with the Rust `dynamo_mocker::replay` implementation behind `python -m dynamo.replay`. That implementation simulates request traces through mock workers and produces performance artifacts. This harness instead feeds recorded `NvCreateChatCompletionStreamResponse` chunks into the in-process HTTP service and observes its engine cancellation context. Commit `117c60120b` updates the module documentation to make that distinction explicit.

On `computelab-build-4`, the final branch state passed all 10 Anthropic replay tests and all 9 Responses replay tests with snapshot updating disabled. I have also selected @PeaBrane for review once the current CI rollup finishes, as requested.
